### PR TITLE
Making it pretty

### PR DESCRIPTION
**Pull request by main--**
btw very nice about yaml is that it supports hex (for device IDs)

### CODE REVIEW

File: windows-gaming-driver/src/config.rs [hunk 1]
```diff
@@ -3,20 +3,124 @@ use std::fs::{File, OpenOptions};
...
 #[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, Copy)]
 pub struct DeviceId {
     pub vendor: u16,
     pub product: u16,
 }
 
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+pub enum UsbBinding {
+    ById(DeviceId),
+    ByPort { bus: u16, port: u16 },
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:17]: Create a new struct `ByPort` or `UsbPort` or so and make the variant `ByPort(UsbPoirt)`. While struct variants in enums are fine for matching, you can't write a function which takes it as argument and you can't use it in a generic context.

File: windows-gaming-driver/src/config.rs [hunk 2]
```diff
@@ -3,20 +3,124 @@ use std::fs::{File, OpenOptions};
...
+    pub frequency: usize,
+    pub format: String,
+    pub channels: usize,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub enum SoundBackend {
+    None,
+
+    Alsa {
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:77]: extract struct

File: windows-gaming-driver/src/config.rs [hunk 3]
```diff
@@ -3,20 +3,124 @@ use std::fs::{File, OpenOptions};
...
+        sink: AlsaSettings,
+        source: AlsaSettings,
+    },
+
+    PulseAudio {
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:82]: dito

File: windows-gaming-driver/src/config.rs [hunk 4]
```diff
@@ -3,20 +3,124 @@ use std::fs::{File, OpenOptions};
...
+    fn default() -> SoundBackend {
+        SoundBackend::None
+    }
+}
+
+#[serde(default)]
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct AlsaSettings {
+    pub name: String,
+
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:100]: interesting newline

File: windows-gaming-driver/src/config.rs [hunk 5]
```diff
@@ -3,20 +3,124 @@ use std::fs::{File, OpenOptions};
...
+    pub unit: AlsaUnit,
+    pub buffer_size: usize,
+    pub period_size: usize,
+}
+
+impl Default for AlsaSettings {
+    fn default() -> AlsaSettings {
+        AlsaSettings {
+            name: "default".to_owned(),
+
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:110]: dito

File: windows-gaming-driver/src/config.rs [hunk 6]
```diff
@@ -82,19 +186,39 @@ impl From<DeviceId> for (u16, u16) {
...
-        let contents = toml::to_string(self).unwrap();
+        let contents = serde_yaml::to_string(self).unwrap();
         let mut file = OpenOptions::new().create(true).write(true)
             .truncate(true).open(path).expect("Failed to open config file");
-        write!(file, "{}", contents).expect("Failed to write config file");
+        writeln!(file, "{}", contents).expect("Failed to write config file");
     }
 
-    pub fn load<P: AsRef<Path>>(path: P) -> Config {
+    pub fn load<P: AsRef<Path>>(path: P) -> Option<Config> {
```
Comment (oberien) [inline windows-gaming-driver/src/config.rs:195]: If implementing config migration, we should do so properly. We will definitely need automatic config migration at some point, so why not do it now? :)

I'd suggest adding a "version" property to the config file. We should first read the version property if possiblea as struct having only that one property, or parse the config file as `Value` and manually extract the "version" information. We should create a new mod containing one function which reads the version and calls all proper upgrade functions, which are in (different) mods: `config/mod.rs` contains all current config stuff. Either `config/version.rs` or `config/version/mod.rs` contains the main upgrade function and either `config/version/version{1,2,…}.rs` or `config/version{1,2,…}.rs` contain the functions to upgrade from the previous to that version.

File: windows-gaming-driver/src/controller.rs [hunk 7]
```diff
@@ -173,3 +188,52 @@ impl Controller {
...
+
+    let mut scanner = iter.scan_devices().unwrap();
+    // FIXME: rust-lang/rust#42222
+    return match scanner.next() {
+        Some(dev) => {
+            let mut bus = None;
+            let mut addr = None;
+            for attr in dev.attributes() {
+                if let Some(val) = attr.value().and_then(OsStr::to_str) {
+                    if attr.name() == "busnum" {
```
Comment (oberien) [inline windows-gaming-driver/src/controller.rs:219]: `match attr.name()`?

File: windows-gaming-driver/src/controller.rs [hunk 8]
```diff
@@ -173,3 +188,52 @@ impl Controller {
...
+        }
+        &UsbBinding::ByPort { bus, port } => {
+            iter.match_attribute("busnum", bus.to_string())?;
+            iter.match_attribute("devpath", port.to_string())?;
+        }
+    }
+
+    let mut scanner = iter.scan_devices().unwrap();
+    // FIXME: rust-lang/rust#42222
+    return match scanner.next() {
```
Comment (oberien) [inline windows-gaming-driver/src/controller.rs:213]: why `return match … { … };`?

**Review by oberien — CHANGES_REQUESTED** [review]
I personally prefer toml over yaml, but /shrug. How does yaml solve that optionals are not shown in the config file in toml? Does it support tuples?

I like the idea of making the sound config part of windows-gaming. Unfortunately that won't make it work on my end ;)

Please update the default config in `misc` to reflect our latest config state with yaml.

File: windows-gaming-driver/src/config.rs (hunk 1 above)
Comment (main--) [reply windows-gaming-driver/src/config.rs:17]: I know that and I don't care. It's fine like this. We never need to pass this around.

File: windows-gaming-driver/src/config.rs (hunk 2 above)
Comment (main--) [reply windows-gaming-driver/src/config.rs:77]: ~

File: windows-gaming-driver/src/config.rs (hunk 3 above)
Comment (main--) [reply windows-gaming-driver/src/config.rs:82]: ~

File: windows-gaming-driver/src/config.rs (hunk 4 above)
Comment (main--) [reply windows-gaming-driver/src/config.rs:100]: The name is conceptually different from all the other things.

File: windows-gaming-driver/src/config.rs (hunk 6 above)
Comment (main--) [reply windows-gaming-driver/src/config.rs:195]: How about no? This sure seems like a smart idea ... until you realize that you basically have to fully duplicate the entire config code for every single version plus conversion code that doesn't even make sense in many corner cases (you had no PCI slots? well yeah good luck finding them).

Honest guess is something around 10x the code for config? And around 20x the work as we can no longer just (de)serialize and be done?

And what do we gain? A user base of virtually no one no longer has to deal with breaking changes to the config of a v0.1 tool.

The proper solution is to just not have any breaking changes. But it doesn't matter right now as we don't have a stable release anyways.

Sure I'm not opposed to doing it like this but I won't be the one developing **and maintaining** it.

File: windows-gaming-driver/src/controller.rs (hunk 8 above)
Comment (main--) [reply windows-gaming-driver/src/controller.rs:213]: literally says that in the line above

File: windows-gaming-driver/src/controller.rs (hunk 7 above)
Comment (main--) [reply windows-gaming-driver/src/controller.rs:219]: doesn't work here, it's an `OsStr`

**Comment by main--** [comment]
>How does yaml solve that optionals are not shown in the config file in toml?

AFAICS it doesn't (kind-of): `setup: ~` is what I got in my config. But I'm fine with that.

>Does it support tuples?

no idea, probably. but we don't need that now anyways

>Please update the default config in misc to reflect our latest config state with yaml.

not sure I'm lazy - only one who needs this is you guys and why bother when I can just tell you / send you mine

In the big picture, you're not supposed to need this anyways. The file in `misc` is a remnant from a time when you were supposed to set this up by creating the config file, wizard wasn't even an idea. We might decide to write a config reference anyways but should do it properly then (fully document).

**Comment by oberien** [comment]
¯\\_(ツ)_/¯

**Review by oberien — APPROVED** [review]

**Review by oberien — CHANGES_REQUESTED** [review]
I do need the structs to be actual structs and not an enum variant ;) At least the `UsbBinding::ByPort` one.

**Pull request merged**